dcurve/src/internal/clr.ts

```
@@ -0,0 +1,151 @@
+
+    // check if the calculation should include payouts (we cannot create a merkle without any contributions)
+    if (grantRoundContributions.totalPot && options.includePayouts) {
+      // construct the payout obj by combining any grants which have the same payout address
+      distribution.distribution.forEach((grantMatch: GrantMatch) => {
+        // each payout address must only appear once in the payoutDistribution
+        let grantIds = [grantMatch.grantId];
+        // check if the payout address is already recorded
+        if (payoutObj[grantMatch.address]) {
+          // get the current set of grantIds
+          const currentGrantIds = payoutObj[grantMatch.address].grantIds;
+          // add the set of grantIds to the arr and filter duplicates
+          grantIds = grantIds.concat(currentGrantIds).filter((item, pos) => currentGrantIds.indexOf(item) === pos);
+        }
+        // ensure the payout address has a match to claim
+        if (grantMatch.match) {
+          // sum the total match for each payout address
+          payoutObj[grantMatch.address] = {
+            grantIds: grantIds,
+            address: grantMatch.address,
+            match: grantMatch.match + (payoutObj[grantMatch.address]?.match || 0),
+          } as PayoutMatch;
+        }
+      });
+      // record the distribution used to generate the merkle
+      distribution.payoutDistribution = Object.values(payoutObj) as PayoutMatch[];
+      // prevent attempting to generate a merkle tree if we're missing the minimum number of leafs
+      if (distribution.payoutDistribution.length > 1) {
+        // generate the merkle tree and record the root
+        distribution.merkle = generateMerkle(distribution.payoutDistribution, grantRoundContributions.currDecimals);
+        distribution.hash = getMerkleRoot(distribution.merkle);
+      } else {
+        // empty state when we can't create a tree
+        distribution.merkleError = 'Missing required context to build tree';
```

Why empty state instead of throwing an error? E.g. if I'm a user do I need to manually check that `merkleError` is undefined to safely use this script? Not saying this is bad, just trying to understand the decision here!

---

`generateMerkle` will throw if its not provided enough data, I wrapped it because we can know ahead of time if the merkle-tree will fail - but I'm happy to back out those checks and let it throw?

---

I'm inclined to let it throw here unless we imagine some DAO UI interface to consume this.

---

And decision here? I'm fine with whatever you all decide! @gdixon @thelostone-mc @corydickson

---

We still need to clarify how we'll consume/expose the merkle/payout distribution, so I'm tempted to leave this in and make that decision in a follow-up PR? @apbendi any thoughts on this?